### PR TITLE
Add time zone config to set non-UTC [databricks]

### PR DESCRIPTION
**Pull request by res-life**
Contributes https://github.com/NVIDIA/spark-rapids/issues/9627

Add time zone config for CI to set and test non-UTC time zone

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/conftest.py
Comment (pxLi) [inline integration_tests/src/main/python/conftest.py:80]: Q: where would this func by used? also pytest conf would relying on run_pyspark script seems weird

can you at least try `os.environ.get('PYSP_TEST_spark_sql_session_timeZone', 'UTC')` to make sure it could have a default?

Comment (res-life) [reply integration_tests/src/main/python/conftest.py:80]: these funcs will be used in pytest xfail. For an existing test case before operator supports non-utc: add xfail(is_non_utc()).

>also pytest conf would relying on run_pyspark script seems weird.

I tested we can get this config in the `conftest`.
conftest is internal file of IT, so it's safe to get Env variable from run_pyspark.

Comment (pxLi) [reply integration_tests/src/main/python/conftest.py:80]: > conftest is internal file of IT, so it's safe to get Env variable from run_pyspark.

hmm OK, for me pytest code itself should at least provide the defaults here

I would like to hear more feedback from other developers~

Comment (res-life) [reply integration_tests/src/main/python/conftest.py:80]: If does not set it, then an error throws:

```
>>> os.environ["PYSP_TEST_spark_sql_session_timeZone"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<frozen os>", line 679, in __getitem__
KeyError: 'PYSP_TEST_spark_sql_session_timeZone'
```

This error will force us to set this cfg.

**Comment by res-life** [comment]
I saw the skip logic in `date_gen.py`: when timezone is not UTC and data gen contains timestamp type, skip the tests. 
pushed a new commit.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
I cherry-picked this PR into https://github.com/NVIDIA/spark-rapids/pull/9482

File: integration_tests/src/main/python/conftest.py
Comment (revans2) [reply integration_tests/src/main/python/conftest.py:80]: I just threw the code together so a get with a default looks like a great addition for safety/robustness.

File: integration_tests/src/main/python/conftest.py
Comment (revans2) [inline integration_tests/src/main/python/conftest.py:83]: I didn't realize `is_tz_utc` existed. I think we should probably only keep one way of checking if the timezone is UTC or not. `is_tz_utc` has the problem that we need a spark session to make it work. That is fine, but it also makes it difficult to use it to skip a test unless it happens after the test starts to run. This patch removes all uses of `is_tz_utc` so perhaps we should also delete the implementation too.

**Review by revans2 — APPROVED** [review]
Just some nits really.

File: integration_tests/src/main/python/conftest.py
Comment (res-life) [reply integration_tests/src/main/python/conftest.py:83]: Done in https://github.com/NVIDIA/spark-rapids/pull/9482
- default value with `PYSP_TEST_spark_sql_session_timeZone`
- Remove useless `is_tz_utc`

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
Building was blocked by https://github.com/NVIDIA/spark-rapids/issues/9681

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
Premerge error:
```
[2023-11-15T08:29:12.100Z] - mixed blocking alloc with spill *** FAILED ***

[2023-11-15T08:29:12.102Z]   java.util.concurrent.TimeoutException:

[2023-11-15T08:29:12.102Z]   at com.nvidia.spark.rapids.HostAllocSuite$TaskThread$TaskThreadTrackingOp.get(HostAllocSuite.scala:107)

[2023-11-15T08:29:12.102Z]   at com.nvidia.spark.rapids.HostAllocSuite$AllocOnAnotherThread.waitForAlloc(HostAllocSuite.scala:218)

[2023-11-15T08:29:12.102Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$50(HostAllocSuite.scala:705)

[2023-11-15T08:29:12.102Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$50$adapted(HostAllocSuite.scala:703)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$49(HostAllocSuite.scala:703)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$49$adapted(HostAllocSuite.scala:699)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$48(HostAllocSuite.scala:699)

[2023-11-15T08:29:12.103Z]   at com.nvidia.spark.rapids.HostAllocSuite.$anonfun$new$48$adapted(HostAllocSuite.scala:690)
```

Get result in 1s but time out exception occurred. Seems it's a ramdom failure.
I'll re-run the premerge.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
@revans2  We need to merge this first.

**Review by revans2 — APPROVED** [review]
But I would like others to review/approve too before merging because I wrote a lot of this code.

File: integration_tests/src/main/python/conftest.py
Comment (jlowe) [reply integration_tests/src/main/python/conftest.py:80]: Agree with @pxLi, this seems weird.  There's already a standard environment variable for users to convey what timezone they want an application to use, and that's `TZ` (see environ(7) manpage).  Is there a reason to avoid using that when running the tests?  Then the user can simply set TZ when they run the tests to specify the timezone to use (e.g.: TZ=CST integration_tests/run_pyspark_from_build.sh ...).  I already do this when running Spark in local mode to get it to use the timezone desired, and it would be nice to avoid adding yet another way to specify a timezone that isn't compatible with the standard way to do it.

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/conftest.py
Comment (res-life) [reply integration_tests/src/main/python/conftest.py:80]: Changed from TEST_TZ to TZ.
Note:  can not get the TZ env variable in conftest.py. So still use  `PYSP_TEST_spark_sql_session_timeZone` to pass time zone.

Comment (jlowe) [reply integration_tests/src/main/python/conftest.py:80]: > can not get the TZ env variable in conftest.py.

Can you elaborate on this?  I'm confused how we can export one variable and read it in conftest but somehow can't do the same to another.  Is something in the shell startup environment bashing the TZ variable?

Comment (winningsix) [reply integration_tests/src/main/python/conftest.py:80]: @res-life wrapping ```TZ``` with ```PYSP_TEST_spark_sql_session_timeZone``` wouldn't be same as original way we did in [is_tz_utc](https://github.com/NVIDIA/spark-rapids/pull/9652/files#diff-0271a6b980cb547f3b0f7a18ab9109357512c6c3562fd3631cbceae3bd3542f4L58)?

Java systemDefault will respect TZ environment.
```java 
import java.time.ZoneId;

public class Test{
   public static void main(String [] args) {
     System.out.println("time zone is " + ZoneId.systemDefault());
   }

}
```
```
$ export TZ="UTC" && java Test
time zone is UTC
$ export TZ="Asia/Shanghai" && java Test
time zone is Asia/Shanghai
```

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/conftest.py
Comment (winningsix) [reply integration_tests/src/main/python/conftest.py:80]: Offline discussed with @res-life , let's add comments that we need to get utc time before spark session starts.

**Comment by winningsix** [comment]
@res-life Per discussed, let's file a ticket tracking DB failed issue if you can produce it with seed ID in last round.

File: integration_tests/src/main/python/conftest.py
Comment (res-life) [reply integration_tests/src/main/python/conftest.py:80]: We can not use the following code when the Spark session is not started.
```python
jvm = spark.sparkContext._jvm
    utc = jvm.java.time.ZoneId.of('UTC').normalized()
    sys_tz = jvm.java.time.ZoneId.systemDefault().normalized()
```

Comment (res-life) [reply integration_tests/src/main/python/conftest.py:80]: >can not get the TZ env variable in conftest.py.

Because TZ is rewrite to UTC in [data_gen.py](https://github.com/NVIDIA/spark-rapids/blob/v23.10.0/integration_tests/src/main/python/data_gen.py#L33-L36)
```
os.environ['TZ'] = 'UTC'
time.tzset()
```

Removed the above code. See my last commit:
I use Env variable TZ now.
And I updated TimestampGen to avoid generate out of range timestamp

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
@jlowe  @NVnavkumar Please help review, thannks.

**Comment by res-life** [comment]
>@res-life Per discussed, let's file a ticket tracking DB failed issue if you can produce it with seed ID in last round.

@winningsix Did not reproduce it in the last premerge.
And I did not reproduce it on DB 3.2.1 using the  seed ID. will try DB 3.3.0

**Review by winningsix — APPROVED** [review]
LGTM

**Comment by res-life** [comment]
@winningsix Also DB 330 dis not reproduce the previous error.

Anyway, record the previous error here
```
[2023-11-16T09:54:48.168Z] [2023-11-16T09:54:09.082Z] =========================== short test summary info ============================
[2023-11-16T09:54:48.168Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[false-0][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[false-1][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[false-2][DATAGEN_SEED=1700125723, INJECT_OOM, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[false-3][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[false-4][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[true-0][DATAGEN_SEED=1700125723, INJECT_OOM, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[true-1][DATAGEN_SEED=1700125723, INJECT_OOM, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[true-2][DATAGEN_SEED=1700125723, INJECT_OOM, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[true-3][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] FAILED ../../src/main/python/delta_zorder_test.py::test_delta_dfp_reuse_broadcast_exchange[true-4][DATAGEN_SEED=1700125723, IGNORE_ORDER({'local': True})] - pyspark.errors.exceptions.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGES...
[2023-11-16T09:54:48.169Z] [2023-11-16T09:54:09.082Z] = 10 failed, 241 passed, 5 skipped, 21949 deselected, 5 xfailed, 9 warnings in 2695.10s (0:44:55) =

```

**Pull request merged**
